pow/ethash.go

```diff
@@ -0,0 +1,581 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pow
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+	"unsafe"
+
+	mmap "github.com/edsrzf/mmap-go"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	metrics "github.com/rcrowley/go-metrics"
+)
+
+var (
+	ErrNonceOutOfRange   = errors.New("nonce out of range")
+	ErrInvalidDifficulty = errors.New("non-positive difficulty")
+	ErrInvalidMixDigest  = errors.New("invalid mix digest")
+	ErrInvalidPoW        = errors.New("pow difficulty invalid")
+)
+
+var (
+	// maxUint256 is a big integer representing 2^256-1
+	maxUint256 = new(big.Int).Exp(big.NewInt(2), big.NewInt(256), big.NewInt(0))
+
+	// sharedEthash is a full instance that can be shared between multiple users.
+	sharedEthash = NewFullEthash("", 3, 0, "", 1, 0)
+
+	// algorithmRevision is the data structure version used for file naming.
+	algorithmRevision = 23
+
+	// dumpMagic is a dataset dump header to sanity check a data dump.
+	dumpMagic = hexutil.MustDecode("0xfee1deadbaddcafe")
+)
+
+// isLittleEndian returns whether the local system is running in little or big
+// endian byte order.
+func isLittleEndian() bool {
+	n := uint32(0x01020304)
+	return *(*byte)(unsafe.Pointer(&n)) == 0x04
+}
+
+// memoryMap tries to memory map a file of uint32s for read only access.
+func memoryMap(path string, write bool) (*os.File, mmap.MMap, []uint32, error) {
+	// Try to open the file and memory map it
+	var (
+		file *os.File
+		mem  mmap.MMap
+		err  error
+	)
+	if write {
+		file, err = os.OpenFile(path, os.O_RDWR, os.ModePerm)
+	} else {
+		file, err = os.OpenFile(path, os.O_RDONLY, os.ModePerm)
```

Comment:
`os.ModePerm` is supposed to be used as a *mask*. Its value is `0777`, all files will be created with read/write/execute permission for all users. It would be better to use a sane mode like `0644`.